designs/2019-relative-plugin-loading/README.md

```diff
@@ -0,0 +1,169 @@
+
+The current system has a few problems:
+
+* It's inconvenient that that the ESLint invoker needs to manually install any plugins required by a shareable config. As a result of this constraint, a shareable config can't add a plugin without requiring new manual installation steps from ESLint invokers. This greatly decreases the ergonomics of using custom rules from plugins in shareable configs, and results in increased pressure to add new rules and options to ESLint core.
+* ESLint loads plugins from the CWD of the process, which is typically somewhere in the end user's project. However, with the current setup the correct place to load plugins from is the ESLint invoker's project. As a result, if the ESLint invoker is not the end user (e.g. in `create-react-app` projects), the default plugin-loading location is incorrect. The ESLint invoker can change the CWD, but changing the CWD has some inconvenient side-effects:
+    * Only the `node_modules` folder in the CWD is ignored by default, so changing the CWD would cause some files to be unignored.
+    * Editor integrations usually won't know to change the CWD (since they might not be able to figure out who the "intended" ESLint invoker is for a given project), so they might fail to load the correct plugins.
+* Users often want to write custom project-specific rules and put them in the project's repository. However, plugins currently need to be in `node_modules`, because allowing other paths for plugins would introduce the possibility of name conflicts.
+
+## Detailed Design
+
+The design has three components:
+
+* Plugins are resolved relative to the config files that depend on them, rather than being loaded relative to the ESLint invoker's project.
+* Plugin name conflicts result in a fatal error.
+* Plugins in configs can be loaded by filepath.
+
+### Plugins are resolved relative to the config files that depend on them
+
+A config is said to "load" a given plugin name if either of the following are true:
+
+* The config includes the plugin name in a `plugins` list, e.g. `plugins: ["foo"]`
+* The config extends a shareable config from the plugin, e.g. `extends: ["plugin:foo/bar"]`
+
+To load a plugin from a config, the following steps are taken:
+
+* If the config loading the name `eslint-plugin-X` is itself a bundled config in plugin called `eslint-plugin-X`, then the load resolves to that plugin itself. (This is called a *self-load*.)
+* Otherwise, the plugin is loaded as a Node module relative to the location of the config file that loads the plugin's name.
+
+**Open question:** Should `extends: ["plugin:foo/bar"]` be considered a "load" of a plugin? (The "Open Questions" section at near the bottom of this RFC contains more details on how this would affect behavior. To understand the tradeoffs, it may be easier to read through this RFC first before considering the open question.)
```

Comment:
No. I don't think `extends` is considered a "load" of a plugin because it doesn't define any rules to be identified to a unique name.

The `extends: ["plugin:foo/bar"]` uses the `configs.bar` property of `eslint-plugin-foo`, but it doesn't register the plugin to the plugin manager. This means it doesn't define any rules and envs.

Reply:
I think there are a few options:

1. Consider `extends: ["plugin:foo/bar"]` to be a "load" of a plugin (and raise an error if `plugins: ["foo"]` is used from a different config file)
1. Don't consider `extends: ["plugin:foo/bar"]` to be a "load" of a plugin. Instead, require `plugins: ["foo"]` to be used somewhere else, and use the already-loaded plugin to resolve `plugin:foo/bar`. (This is the option described in the ["Open Questions" section](https://github.com/eslint/rfcs/blob/5ea5c19339deff933b87326c4d29435c3c0e3167/designs/2019-relative-plugin-loading/README.md#should-extends-pluginfoobar-be-considered-a-load-of-a-plugin).)
1. Don't consider `extends: ["plugin:foo/bar"]` to be a "load" of a plugin. However, still resolve `extends: ["plugin:foo/bar"]` from the location of the config file that contains it, even if `plugins: ["foo"]` appears in a different config and resolves to a different plugin.

Personally, I think (2) is the best option.

(3) is logically consistent, because `extends: ["plugin:foo/bar"]` doesn't introduce any rules/envs. However, I think it could be confusing in practice, because:
* Usually, a config from a plugin depends on the plugin's rules. So the plugin config contains a self-load with `plugins: ["foo"]`, it would create a conflict anyway.
* If the plugin config doesn't contain `plugins: ["foo"]`, then the config from the plugin could be applied to rules from a different version of the plugin, potentially causing confusing schema errors.
* Users might be surprised that a plugin's shareable configs are from a different version of the plugin than the plugin's rules.

Reply:
Hm. I prefer the **1** in those options because:

1. `extends:"plugin:foo/bar"` is enough clear to use the plugin `foo`.
2. people is familiar with such a behavior because it's the popular usage that  the `recommended` preset contains `plugins` setting.

Reply:
I think the advantage of (2) over (1) would be that it allows more valid configurations. However, let's discuss https://github.com/eslint/rfcs/pull/14#discussion_r260934018 first, since if we implement that suggestion, it would be unnecessary to use (2).

Reply:
Now that the discussion at https://github.com/eslint/rfcs/pull/14#discussion_r260934018 is resolved:

I think (2) is better because users occasionally might want to use `extends: [plugin:foo/bar]` for a shareable config's plugin, and if they can't use `extends: [plugin:foo/bar]` then it's sometimes difficult to make their config do what they want. On the other hand, if we make the change then it's easy for a user to restore the old behavior by simply adding `plugins: ["foo"]` to their config, and it would allow us to support `extends: [plugin:foo/bar]` for a shareable config's plugin.

Reply:
I'm still feeling verbose if we require both `extends: "plugin:foo/recommended"` and `plugins: ["foo"]`. I guess many people felt the same, so most of plugins have `plugins: ["foo"]` in their recommended config. Therefore, my preference is to keep compatibility with this; people can use plugin's config without `plugins` property. (or we remove `plugin:` prefix?)

Reply:
I tried to abandon my adherence to option 1, but I failed by some technical issues.

Option 2 doesn't fit #13. Option 2 means that we cannot determine `extends` until it makes the final config for each file because we can write `plugins` setting in `overrides`. So we cannot build `ConfigArray` under the option.
In #13, I hope to make the order that it loads configs fully at first then determines the config for each file without fs. Probably I have to stick at option 1 or 3.

Reply:
In order to allow https://github.com/eslint/rfcs/pull/10 to work, plugin names need to be globally unique across all files. It's not possible to have one version of `eslint-plugin-foo` loaded for `*.js` and another version of `eslint-plugin-foo` loaded for `*.ts`, because formatters need to obtain unique metadata for the rule `foo/bar`. So I think we'll be able to resolve plugin configs before determining the final config for each file anyway.

We could accomplish that by "hoisting" all `plugins` directives that appear in `overrides`. For example:

```js
module.exports = {
    ...base,
    overrides: {
        ...overridesParams,
        plugins: ["foo"]
    }
};
```

would be equivalent to this config:

```js
module.exports = {
    ...base,
    plugins: ["foo"],
    overrides: {
        ...overridesParams
    }
};
```

Reply:
Thank you for pointing out. I had overlooked the impact of #13 for #10.

Reply:
Given this new information, do you think Option 2 from https://github.com/eslint/rfcs/pull/14#discussion_r260934656 is feasible? If so, I want to update this RFC so that we can have a final version before Thursday's TSC meeting.

Reply:
It may be feasible, but I guess that it needs additional changes regardless of #13.

- Without #13, it loads config files from the leaf to the root on directories. We have to change this order with keeping the `root` property functionality. This might be tough because plugin's config can have `root` property.
- With #13, as same, it needs to change the order for the ancestors of CWD. And it needs a parameter that is the loaded plugin pool to load because it cannot access the config array elements which have been yielded in the past on loading. This may make a pro of #13, the removal of registrations, weak.

Therefore, I'm still prefer the option 1 or 3. That a loading logic for each config file doesn't depend on other configs makes the loading logic simpler much.

---

By the way, if we go with 2, how about the mix of 1 and 2? It looks to cohabit backward compatibility and increasing valid configs.

- If the relevant plugin has been loaded already, it uses the loaded plugin instance. Otherwise, it considers `extends: ["plugin:foo/bar"]` to be a "load" of a plugin (and raise an error if `plugins: ["foo"]` is used from a different config file).